Preliminary Ruby 3.0 support

johnnyshields opened the pull request:
Fixes MONGOID-5048, MONGOID-4974, MONGOID-5041, MONGOID-4833, MONGOID-5044

Summary:
- Allow Ruby 3.0 in tests
- I18n.translate should use double-splat keyword arguments
- Upgrade Rspec to 3.10+ for Ruby 3.0 support (tests fail otherwise)
- Remove `rspec-mocks-diag` gem and switch back to standard `rspec-mocks` (reverts https://github.com/mongodb/mongoid/pull/4690)

Specs seem to be running fine on my local machine with Ruby 3.0 (* test run still in progress but no failures yet). I am not familiar with Evergreen CI and I need some help to get it to run.

p-mongo commented on lib/mongoid/errors/mongoid_error.rb:
```diff
@@ -48,7 +48,7 @@ def compose_message(key, attributes = {})
       # @return [ String ] A localized error message string.
       def translate(key, options)
-        ::I18n.translate("#{BASE_KEY}.#{key}", options)
+        ::I18n.translate("#{BASE_KEY}.#{key}", **options)
```
My thinking with Ruby 3 support is that the double-splat syntax should be used everywhere for options.

Therefore, for example, this method's signature should be changed to take **options, and so on up the call stack.

johnnyshields replied:
@p-mongo lets do that as a separate PR. The focus of this PR should be to get Ruby 3.0 to pass the tests with the minimal amount of code changes.

p-mongo replied:
We generally leave it up to the commenters to resolve conversations on their comments. This helps with the review process as the responses are visible.

I generally try not to add hacks to Mongoid (or driver, etc.) because later on they end up causing larger problems. Ruby 3 uses the double splat syntax for options, therefore it is my expectation that a Ruby 3-compatible release of Mongoid would do the same unless there is a clear reason not to.

johnnyshields replied:
Sure, you're welcome to add this for the "official Ruby 3" release. I have renamed this PR as "Preliminary Ruby 3.0 support" in order to clarify that I am simply getting the tests to pass, so that people who want to use an alpha version on Ruby 3.0 don't have to fork the official repo. I will leave it up to the MongoDB team to add whatever they feel is needed for official Ruby 3.0 support before they publicly announce it. Sound fair?

johnnyshields replied:
Resolving this as per above comment.

johnnyshields commented:
@p-mongo lets do that as a separate PR. The focus of this PR should be to get Ruby 3.0 to pass the tests with the minimal amount of code changes.

alcaeus commented:
@johnnyshields thanks for starting this process! Unfortunately it looks like a number 3.0 and 2.7 variants are failing now. For example, the 3.0 test fails related to `I18n::translate`: https://evergreen.mongodb.com/task_log_raw/mongoid_drivers_ruby_3.0__mongodb_version~4.4_topology~standalone_ruby~ruby_3.0_driver~current_test_patch_1fe51ac7bfabdcf7a406b7a1b94f77ea6a2c6261_60afce1b57e85a6f75bb6eef_21_05_27_17_11_34/0?type=T#L628

@comandeo a couple of builds seem to be breaking on driver-master, it looks like our oppressive terminology projects broke Mongoid due to the removal of some constants around legacy hello: https://evergreen.mongodb.com/task_log_raw/mongoid_driver_master_old__mongodb_version~2.6_topology~standalone_ruby~ruby_2.5_driver~master_test_patch_1fe51ac7bfabdcf7a406b7a1b94f77ea6a2c6261_60afce1b57e85a6f75bb6eef_21_05_27_17_11_34/0?type=T#L570

johnnyshields commented:
@alcaeus how can I access Evergreen? I am a frequent contribute to MongoDB's Ruby projects so if it's possible to get me an account it would be great.

alcaeus commented:
> @alcaeus how can I access Evergreen? I am a frequent contributor to MongoDB's Ruby projects so if it's possible to get me an account it would be great.

Oh, I'm sorry, I forgot that you can't see test failures. I'm not sure if external contributors can get access to logs, maybe @p-mongo knows more about that. I've copied the relevant output in a gist: https://gist.github.com/alcaeus/76d81e8aa55877b17c705c15ef52789e

johnnyshields commented:
@alcaeus I discovered that RSpec was silently causing issues. I have removed the `rspec-mocks-diag` gem, and replaced it with standard `rspec-mocks`. @p-mongo what was the reason for `rspec-mocks-diag`?

alcaeus commented:
Sure thing! Here's the log for ruby 3.0 driver-stable (as before): https://gist.github.com/alcaeus/a2b4bca483f05d13fa0dc5b22249fb28. This time I skipped removing extra output and have left that as an exercise for the reader. @johnnyshields @p-mongo how much effort do you think would it take to run a single variant (driver-stable, ruby 3.0, standalone server) on GitHub Actions to give outside contributors a rough indication of whether things are fine (and we should run our full suite on evergreen) or if the baseline already fails.

johnnyshields commented:
@alcaeus github actions should not be hard, I've set it up before. Will take a look at it tomorrow.

johnnyshields commented:
@alcaeus I believe I have all tests passing, at least on my local. When I ran the entire suite, I got 7 failures. Of these 4 also failed on Ruby 2.7, and the remaining 3 passed reliably when I re-ran them (they would also sometimes fail on Ruby 2.7). In summary, I see no failures which appear to specific to Ruby 3.0.

![image](https://user-images.githubusercontent.com/27655/120072902-eef9a200-c0d0-11eb-91c8-a303b05dae6c.png)

johnnyshields commented:
As for github actions, since it requires an admin for the repo, it's best if someone on MongoDB side can set it up. There are tons of example files in other repos you can copy; it should be pretty easy.

alcaeus commented:
There are a large number of failures in the driver-master builds due to other projects, so I'd say we wait until we have those fixed in the drivers and then take another look at the build results.

As for GitHub actions, I created RUBY-2655 to track this on our end and I'll try to get that squeezed in somewhere (or attempt to do this myself since I've previously set up GitHub Actions for a couple of other drivers)

johnnyshields commented:
@alcaeus I believe I've done everything I can on this PR--again all tests are passing for me. It would be good if MongoDB team could own this PR from now.

reidmorrison commented on lib/mongoid/errors/mongoid_error.rb:
```diff
@@ -48,7 +48,7 @@ def compose_message(key, attributes = {})
       # @return [ String ] A localized error message string.
       def translate(key, options)
-        ::I18n.translate("#{BASE_KEY}.#{key}", options)
+        ::I18n.translate("#{BASE_KEY}.#{key}", **options)
```
:+1: :tada:

comandeo commented:
@johnnyshields Could you allow pushing your repository? Thank you!

johnnyshields commented:
@comandeo Github is not letting me enable that for some reason.

Can you close this PR and open a new one?

comandeo commented:
Closing this one because of permission issued. Ruby 3.0 support work will be continued here - https://github.com/mongodb/mongoid/pull/5004
